test/make_dist.py

```diff
@@ -47,5 +50,104 @@ def make_dist(devel_mode):
+    except subprocess.CalledProcessError:
+        print("make_dist: not a git repository")
+        return None
+
+    dists = glob.glob(f"cockpit-*{sha[:8]}*.tar.xz")
+    if dists:
+        print("make_dist: already downloaded " + dists[0])
+        return os.path.abspath(dists[0])
+
+    sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "bots"))
+    import task
+
+    # task.api defaults to the current checkout's origin, but the artifacts are on cockpit-project
+    # except if a developer wants to change the artifact building workflow on their fork, support that with $GITHUB_BASE
+    api = task.github.GitHub(repo=os.getenv("GITHUB_BASE", "cockpit-project/cockpit"))
+
+    # downloading GitHub artifacts requires a token
+    if not api.token:
+        print("make_dist: no GitHub API token available")
+        return None
+
+    # iterate artifacts and search our SHA
+    page = 1
+    batch_size = 100
+    download_url = None
+    while not download_url and batch_size == 100:
+        batch = api.get(f"actions/artifacts?per_page={batch_size}&page={page}")["artifacts"]
+        for artifact in batch:
+            if artifact["name"] == "dist-" + sha:
+                download_url = artifact["archive_download_url"]
+                break
+        # if the current batch is < 100, we have the last page and can stop
+        batch_size = len(batch)
+        page += 1
+
+    if not download_url:
+        print(f"make_dist: no download available for commit {sha}")
+        return None
+
+    print(f"make_dist: Downloading dist tarball from {download_url} ...")
+    request = urllib.request.Request(download_url, headers={"Authorization": "token " + api.token})
+    zipio = io.BytesIO()
+    try:
+        with urllib.request.urlopen(request) as response:
+            if os.isatty(sys.stdout.fileno()):
+                total_size = 0
+            else:
+                total_size = None
+            MB = 2**20
+            # read zip into a stringio, as the stream is not seekable and zip requires that
+            while True:
+                block = response.read(MB)
+                if len(block) == 0:
+                    break
+                if total_size is not None:
+                    total_size += len(block)
+                    sys.stdout.write(f"\r{ total_size // MB } MB")
+
+                zipio.write(block)
+
+            # clear the download progress in tty mode
+            if total_size is not None:
+                sys.stdout.write("\r                             \r")
+
+    except urllib.error.HTTPError as e:
+        print("make_dist: Download failed:", e)
+        return None
+
+    with zipfile.ZipFile(zipio) as fzip:
+        names = fzip.namelist()
+        if len(names) != 1 or not names[0].endswith(".tar.xz"):
+            print("make_dist: expected zip artifact with exactly one tar.xz member")
```

Comment:
Hmm was this for your debugging or it can really happen?

Reply:
It Should Not Happen™ right now, but if we ever change our workflow to e.g. include cockpit-cache, this will then tell us why the download is not working (i.e. we need to adjust make_dist to the new tarball structure)